### ISSUE TITLE
How to correctly add extra trainable module in qwen-2.5-vl?

### ISSUE THREAD

**Issue by JerryPW** [QwenLM/Qwen3-VL#1646]
Sorry to bother. My question is more of an engineering issue rather than a bug in Qwen itself. Since I’m still a beginner and not very familiar with MLLM code, I’ve spent several days debugging without much progress. So this post is meant to be a personal technical help request, not a bug report about Qwen.

I want to extract the image tokens from the mid layers of the Qwen-2.5-VL-7B model and project their dimensions to match those of DINOv3, in order to implement a simple visual alignment method. Since this requires modifying the model architecture, I copied the Qwen2_5_VLForConditionalGeneration class from the Transformers library and rewrote it myself, adding an alignment projector to map the image token hidden size to the DINOv3 feature dimension, as shown below:
```
class Qwen2_5_VL_VRA(Qwen2_5_VLPreTrainedModel, GenerationMixin):
    _checkpoint_conversion_mapping = {
        "Visual": "model.visual",
        r"^model\.(?!.*(language_model|visual))": "model.language_model",
    }

    _tied_weights_keys = ["lm_head.weight"]
    # Reference: fix gemma3 grad acc #37208
    accepts_loss_kwargs = False

    def __init__(self, config):
        super().__init__(config)
        ### This is the code from the original Qwen2_5_VLForConditionalGenration class
        self.model = Qwen2_5_VLModel(config)
        self.lm_head = nn.Linear(config.text_config.hidden_size, config.text_config.vocab_size, bias=False)
        ###
        
        # ...some configuraton code
        self.alignment_projector = AlignmentProjector(config.hidden_size, self.projector_dim, self.z_dim)

        if "dino" in self.vra_target:
            from transformers import AutoModel, AutoImageProcessor
            self.dino_processor = AutoImageProcessor.from_pretrained(config.vra_model_path)
            self.alignment_encoder = AutoModel.from_pretrained(config.vra_model_path)

        self.post_init()
```

The fine-tuning code was directly copied from ```qwen-vl-finetune/qwenvl/train/train_qwen.py```. The only change I made was replacing ```Qwen2_5_VLForConditionalGeneration.from_pretrain``` with ```Qwen2_5_VL_VRA.from_pretrain```; everything else remained exactly the same.

However, the problem is that when I use this alignment projector for projection, it maps my features into an all-ones vector. What’s even more puzzling is that when I inspect the parameter weights of the alignment projector using the following code, it often outputs empty tensors with size 0.
```
rank0_print(self.alignment_projector[0])
for name, p in self.alignment_projector[0].named_parameters():
     rank0_print(f"{name}  shape={tuple(p.shape)}")
     rank0_print(p.detach().cpu())
```
printed outcome: 
```
AlignmentProjector(
  (fc1): Linear(in_features=3584, out_features=2048, bias=True)
  (act1): SiLU()
  (fc2): Linear(in_features=2048, out_features=2048, bias=True)
  (act2): SiLU()
  (fc3): Linear(in_features=2048, out_features=768, bias=True)
)
fc1.weight  shape=(0,)
tensor([])
fc1.bias  shape=(0,)
tensor([])
fc2.weight  shape=(0,)
tensor([])
fc2.bias  shape=(0,)
tensor([])
fc3.weight  shape=(0,)
tensor([])
fc3.bias  shape=(0,)
tensor([])
```
However, when I print the parameters of each layer inside the AlignmentProjector, the results look like the ones shown below. For each linear layer, its weights are all zeros (w = 0) and biases are ones (b = 1), which clearly explains why my projector outputs all-one feature vectors.

But my first question is: why can’t I correctly see my parameters using the printing method above? Normally, when I run similar code locally, I can view the parameter values without any issue.
```
def forward(self, x):
    xt = x.detach().float()
    rank0_print("x ", tuple(xt.shape), xt.dtype, xt.min().item(), xt.max().item(),
                "unique-", torch.unique(xt).numel())

    w, b = self.fc1.weight, self.fc1.bias
    rank0_print("w1", w.dtype, "max|w|=", w.detach().abs().max().item(),
                "allzero=", bool((w.detach() == 0).all()))
    rank0_print("b1", b.dtype, "min/max b=", b.detach().min().item(), b.detach().max().item(),
                "allzero=", bool((b.detach() == 0).all()))

    # ... the rest of function forward

```
```
x (8, 646, 3584) torch.float32 -7552.0 1968.0 unique= 4056
w1 torch.bfloat16 max|w|= 0.0 allzero= True
b1 torch.bfloat16 min/max b= -1.0 1.0 allzero= False

```
The remaining issue is how to fix the all-ones output problem. When loading Qwen-2.5-VL-7B, the code gives a warning that the parameters of the alignment projector are “unexpected keys.” However, since the entire training process is wrapped inside a simple Trainer.train() call, I’m not sure how parameter initialization is handled internally. I also don’t understand why, even though I set requires_grad=True for the alignment projector, its parameters remain at 0 and 1 instead of being updated during training.

So my question is: if I want to add a new module (like the alignment projector) to the original Qwen model, how should I properly initialize its parameters and include them in the trainer’s optimizer so that they can be trained correctly with gradients?

My personal guess is that the issue might be related to using DeepSpeed for distributed training, which could make some of the code written for single-GPU inference behave incorrectly. However, since everything is wrapped inside trainer.train(), it’s extremely complicated to inspect the details one by one.

Thank you for reading this far. I would be deeply grateful if you could provide any effective help.

**Comment by JerryPW** [comment]
The issue turned out to be that the learning rate was too small, which made the parameters barely trainable.
Never trust the official fine-tuning documentation that suggests a learning rate of 2e-7!!!!!!!!!!
At least I’ve learned something — for SFT, a learning rate in the range of 1e-5 to 1e-4 is usually much more appropriate.

**Comment by John-Ge** [comment]
If you want to finetune the model to align with some tasks not exist in pretrain stage, you should set a higher learning rate. If you do not want to harm the pretrain capability, you should set a smaller learning rate.

**Comment by JJJYmmm** [comment]
usually, you should solely train the alignment projector with large lr.
zeros and ones are the default value for nn.Linear, you can use other normalization methods like kaiming_norm
